gcc/rust/backend/rust-compile-expr.cc

```diff
@@ -803,7 +803,23 @@ CompileExpr::visit (HIR::WhileLoopExpr &expr)
   ctx->add_statement (loop_begin_label_decl);
   ctx->push_loop_begin_label (loop_begin_label);
 
-  tree condition = CompileExpr::Compile (expr.get_predicate_expr (), ctx);
+  HIR::Expr &predicate = expr.get_predicate_expr ();
+  TyTy::BaseType *predicate_type = nullptr;
+  bool ok
+    = ctx->get_tyctx ()->lookup_type (predicate.get_mappings ().get_hirid (),
+				      &predicate_type);
+  rust_assert (ok && predicate_type != nullptr);
+  tree condition;
+  condition = CompileExpr::Compile (predicate, ctx);
+  if (condition == error_mark_node)
```

Comment:
you can just ignore the case of condition == error mark node... because it will show up in -fdump-tree-gimple, not sure how it could error something else should spit out and error message so i would just do:

```
tree condition = CompileExpr::Compile (predicate, ctx);
if (predicate_type->get_kind() == TyTy::TypeKind::NEVER)
{
    // The predicate was compiled and added to the tree.
    // For NEVER type, we need to ensure it's in the output, so explicitly add it
    ctx->add_statement(condition);
    condition = boolean_true_node;
}
```